external audit: proto spec

tobiaszheller opened the pull request:
This PR adds proto spec for service that will be used in external audit/byob feature.
Service will be implemented in enterprise repo.

Documentation is added as part of proto.


Part of: https://github.com/gravitational/cloud/issues/4661
RFD: https://github.com/gravitational/cloud/blob/rfd/77-bring-your-own-bucket/rfd/0077-Bring-your-own-bucket.md

marcoandredinis commented:
What's the expected flow?
If I have the ExternalAudit configured and deployed, what happens when I update it? Does it keep running using the previous version until I run Ensure again?

What if I delete it?

Can we update all the fields of the ExternalAudit after it is deployed?

I think we are missing some docs on how the service is going to work
If this is documented somewhere else, please ignore my comments :pray: 

The test connection specification shouldn't be required here, I think
You should be able to re-use this:
https://github.com/gravitational/teleport/blob/77c11a3a4e01941d92b2d17a6ed1c4341fce0eee/lib/web/apiserver.go#L784-L785

Commit:
externalaudit: proto spec

tobiaszheller commented:
> What's the expected flow? If I have the ExternalAudit configured and deployed, what happens when I update it? Does it keep running using the previous version until I run Ensure again?
> 
> What if I delete it?
> 
> Can we update all the fields of the ExternalAudit after it is deployed?
> 
> I think we are missing some docs on how the service is going to work If this is documented somewhere else, please ignore my comments 🙏
> 
> The test connection specification shouldn't be required here, I think You should be able to re-use this:
> 
> https://github.com/gravitational/teleport/blob/77c11a3a4e01941d92b2d17a6ed1c4341fce0eee/lib/web/apiserver.go#L784-L785

@marcoandredinis finally I have chosen version with status on each external_audit resource. I think it's the most clean one.

marcoandredinis commented on api/proto/teleport/externalaudit/v1/externalaudit.proto:
```diff
@@ -0,0 +1,66 @@
+// ExternalAudit contains external audit configuration.
+// It's used only in Teleport Cloud with feature called "bring your own bucket".
+// It contains configuration that allows store audit events and session
+// recordings on customer infra instread of teleport Cloud.
```
```suggestion
// recordings on customer infra instead of Teleport Cloud.
```

marcoandredinis commented on api/proto/teleport/externalaudit/v1/externalaudit_service.proto:
```diff
@@ -0,0 +1,90 @@
+  // First you need to SetExternalAuditStatus to `draft` on that resource.
+  rpc DeleteExternalAudit(DeleteExternalAuditRequest) returns (google.protobuf.Empty);
+
+  // SetExternalAuditStatus sets external audit status based on name of external audit.
+  // Setting to draft->active or active->draft will cause service reload.
+  // Only single resource can be set to active state.
+  rpc SetExternalAuditStatus(SetExternalAuditStatusRequest) returns (SetExternalAuditStatusResponse);
```
Can we change between configs?
That means we would get 0 active or 2 active, which does not make sense I think.
That's why I was suggesting to have this resource being referenced by another global setting

tobiaszheller replied:
I have removed that endpoint + status and added new `SetClusterExternalAudit`. It should make more sense now.

marcoandredinis commented on api/proto/teleport/externalaudit/v1/externalaudit_service.proto:
```diff
@@ -0,0 +1,90 @@
+//
+// Supporting multiple ExternalAudit as resources allow us to to add resource,
+// and test connection before actually enabling it in cluster.
+service ExternalAuditService {
```
When are we executing the infra bootstrap?

tobiaszheller replied:
I think it will passed either as a flag to `CreateExternalAudit` or separate RPC endpoint. Maybe separate endpoint makes most sense.

marcoandredinis replied:
So, is it a separate endpoint?

tobiaszheller replied:
yep, i think so, bootstrap can also take a while (up to few seconds - although I haven't measure it yet), so it makes sense to keep is as separate endpoint.

marcoandredinis replied:
I assume we are going to add it to a separate PR

tobiaszheller replied:
yep, exactly, sorry for not being specific

tobiaszheller commented:
@hugoShaka @tcsc it should be ready to review now

tobiaszheller commented:
@tcsc @marcoandredinis friendly ping, it should be ready to review

marcoandredinis commented on api/proto/teleport/externalaudit/v1/externalaudit_service.proto:
```diff
@@ -0,0 +1,100 @@
+//
+// Supporting multiple ExternalAudit as resources allow us to to add resource,
+// and test connection before actually enabling it in cluster.
+service ExternalAuditService {
```
I'm not seeing a Listing operation.
I think we might need one, although I'm not sure.
Is there a case where we will be listing existing Configs?

tobiaszheller replied:
i don't think we will need it, we can always add it later when it's needed

Commit:
Add ClusterExternalAudit to event.proto

espadolini commented on api/proto/teleport/externalaudit/v1/externalaudit.proto:
```diff
@@ -0,0 +1,68 @@
+// It's used only in Teleport Cloud with feature called "bring your own bucket".
+// It contains configuration that allows store audit events and session
+// recordings on customer infra instead of Teleport Cloud.
+message ExternalAudit {
```
The name `ExternalAudit` is far too generic for it to mean "configuration for athenaevents and s3sessions using the cloud-specific OIDC integration for aws credentials". Maybe `ExternalCloudAudit`? Is this going to be in a service that's only spun up in Teleport Cloud?

espadolini commented on api/proto/teleport/externalaudit/v1/externalaudit_service.proto:
```diff
@@ -0,0 +1,100 @@
+
+// SetClusterExternalAuditRequest is a request for setting an cluster external audit status.
+message SetClusterExternalAuditRequest {
+  // ClusterExternalAudit contains name of external audit to set as cluster external audit.
+  ClusterExternalAudit cluster_external_audit = 1;
```
If we only need the name here, why not just a `string name`, like for the Delete operation?

espadolini commented on api/proto/teleport/externalaudit/v1/externalaudit_service.proto:
```diff
@@ -0,0 +1,100 @@
+  // name of external audit.
+  // This method cause Teleport Auth service reload.
+  // Only single resource can be set to active state.
+  rpc SetClusterExternalAudit(SetClusterExternalAuditRequest) returns (SetClusterExternalAuditResponse);
```
What do you think of `EnableClusterExternalAudit` and `DisableClusterExternalAudit` rather than `Set`/`Delete`? `Set` might give the impression that it's just some configuration operation, not that it will cause some whole-cluster disruption.

espadolini commented on api/proto/teleport/externalaudit/v1/externalaudit_service.proto:
```diff
@@ -0,0 +1,100 @@
+service ExternalAuditService {
+  // GetExternalAudit returns the external audit configuration resource.
+  rpc GetExternalAudit(GetExternalAuditRequest) returns (GetExternalAuditResponse);
+
```
You're missing a `GetExternalAudits` or `ListExternalAudits` operation of some sorts.

espadolini commented on api/proto/teleport/externalaudit/v1/externalaudit_service.proto:
```diff
@@ -0,0 +1,100 @@
+
+  // CreateExternalAudit creates a new External Audit resource.
+  rpc CreateExternalAudit(CreateExternalAuditRequest) returns (CreateExternalAuditResponse);
+
```
The absence of any Update/Upsert/ConditionalUpdate/CompareAndSwap operation is deliberate so we never have to handle the case of updating an external audit that's currently in use, right? I'd make a note with a comment about it.

tobiaszheller replied:
yep, that's intentional, so someone won't break existing config used in cluster. Bascially it should be not mutable and if someone want to "update their existing config", new external audit resorce should be created, we can test connection to it and later enable it. 
Thanks for pointing missing docs, I will update it.

Commit:
comments

Commit:
rework to externalcloudaudit

tobiaszheller commented:
@espadolini @marcoandredinis
I have update proto PR with following changes:
- rename `external_audit` to `external_cloud_audit`
- added description why update external_cloud_audit is missing
- renamed Set/DeleteClusterExternalAudit to Enable/DisableClusterExternalCloudAudit
- Added missing `GetExternalCloudAudits` and `GetClusterExternalCloudAudit`

nklaassen reviewed (commented):
Edit: disregard this, I thought better of it

I had some ideas. It seems to me there will be 3 stages to enabling this, each needing its own RPC

(Not counting Step 0. Update OIDC role with permissions to bootstrap infra and view events/recordings)

## Step 1. Bootstrap infra

```
rpc BootstrapExternalCloudAuditInfra(BootstrapExternalCloudAuditInfraRequest) returns BootstrapExternalCloudAuditInfraResponse
```

The request will include the existing OIDC integration name and a name for the new `ExternalCloudAudit` config resource. I think the implementation of this RPC should create all the infra and also should create the `ExternalCloudAudit` resource. Maybe it could actually be a streaming endpoint to report incremental progress. It should create a new mostly empty `ExternalCloudAudit` resource as soon as it starts and incrementally write the name of each created resource as it goes. This can help with cleanup and picking up where it left off if it gets interrupted. It should be able to automatically resume on subsequent calls if it gets interrupted.

## Step 2. Test the configuration

```
rpc TestExternalCloudAudit(TestExternalCloudAuditRequest) returns TestExternalCloudAuditResponse
```

The request should include the name of the ExternalCloudAudit config to test, and the implementation should write the test results to the ExternalCloudAudit resource as well as returning them, so that past test results can be viewed, and we can block you from enabling an ExternalCloudAudit without a successful test.

## Step 3. Enable the configuration

```
rpc EnableExternalCloudAudit(EnableExternalCloudAuditRequest) returns EnableExternalCloudAuditResponse
```

The request should include the name of the ExternalCloudAudit config to enable. It should have all resources populated and it should have a passing test run included. This turns on the feature and triggers auth restarts.

## Possible Step 4. Cleanup bootstrapped infra

```
rpc CleanupExternalCloudAuditInfra(CleanupExternalCloudAuditInfraRequest) returns CleanupExternalCloudAuditInfraResponse
```

The request should include the name of the `ExternalCloudAudit` config. This attempts to clean up all the infra we created in the customer AWS account. The `ExternalCloudAudit` must not be enabled.

---

Note: I don't think there should be any exposed RPCs to create or modify the `ExternalCloudAudit` resource, creation happens in `BootstrapExternalCloudAuditInfra` and some updates happen there as well as in `TestExternalCloudAudit` but they are not exposed to users. We could add a create endpoint in the future if we allow customers to create the infra on their own. There can still be a Delete rpc.

---

Sorry this is on the proto PR and I wasn't here for the RFD review, but wondering if these ideas make sense? To me it makes it clear where all the resource name/uri fields get populated and it addresses some concerns of where we can hold state during resource bootstrapping, how to force users to test and allow them to retest, and how to cleanup.

tobiaszheller commented:
> ## Step 1. Bootstrap infra
> ## Step 2. Test the configuration
> ## Step 3. Enable the configuration
> ## Possible Step 4. Cleanup bootstrapped infra

In MVP of that feature we want to bootstrap infra, but in [next phase](https://github.com/gravitational/cloud/pull/5198#pullrequestreview-1598710903) we want to allow customers to configure bucket on their own and just pass address of it. I think we should prepare our API with that in mind. If we don't enforce bootstrap of infra, doing phase 2 will be trivial (we just need to provide them docs/sample with permissions).
That's why I was thinking that step 1. should be creating resource, step 2 (optional) bootstrapping of infra, step 3 test connection, step 4 enabling. 

As far I understand your description, we will need to have additional logic that enforce that only single external audit resource is active, right? 
@nklaassen

nklaassen reviewed (approved):
Thought about it some more, forget what I was saying earlier, it should be fine to decide all the resource names up front and create the resource first before bootstrapping. This resource CRUD api will be good to get in now, we can add the bootstrap/test RPCs later.

tobiaszheller commented:
closing in favor of https://github.com/gravitational/teleport/pull/33022
